### PR TITLE
referral tracker

### PR DESCRIPTION
**Pull request by megaherz**

### CODE REVIEW

File: cmd/referral/main.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	BlockchainMainKeyringBackend     string `long:"blockchain.main.keyring_backend" env:"BLOCKCHAIN_MAIN_KEYRING_BACKEND" default:"test" description:"decentrcli keyring backend"`
+	BlockchainMainKeyringPromptInput string `long:"blockchain.main.keyring_prompt_input" env:"BLOCKCHAIN_MAIN_KEYRING_PROMPT_INPUT" description:"decentrcli keyring prompt input"`
+	BlockchainMainGas                uint64 `long:"blockchain.main.gas" env:"BLOCKCHAIN_MAIN_GAS" default:"10" description:"gas amount"`
+	BlockchainMainFee                string `long:"blockchain.main.fee" env:"BLOCKCHAIN_MAIN_FEE" default:"1udec" description:"transaction fee"`
+	BlockchainMainRESTNodeURL        string `long:"blockchain.main.rest_node_url" env:"BLOCKCHAIN_MAIN_REST_NODE_URL" default:"http://hera.mainnet.decentr.xyz" description:"REST endpoint URL"`
+
+	ReferralSenderReward   int `long:"referral.sender_reward" env:"REFERRAL_SENDER_REWARD" description:"referral sender reward uDEC'"`
+	ReferralReceiverReward int `long:"referral.receiver_reward" env:"REFERRAL_RECEIVER_REWARD" description:"referral receiver reward uDEC'"`
```
Comment (mikluke) [inline cmd/referral/main.go:52]: maybe it's better to use sdk.Coin or sdk.Int

File: internal/referral/rewarder.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	go func(ticker *time.Ticker) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				r.do(ctx)
+			}
+		}
+	}(ticker)
```
Comment (mikluke) [inline internal/referral/rewarder.go:53]: shouldn't we run `do`  at start as well?

File: internal/referral/rewarder.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+			"installed at":    ref.InstalledAt,
+		})
+
+		resp, err := r.brc.GetTokenBalance(ctx, ref.Receiver)
+		if err != nil {
+			logger.WithError(err).Error("failed to get PDV token balance")
+			continue
+		}
+
+		PDVBalanceFloat64, err := strconv.ParseFloat(resp.Result.Balance, 64)
```
Comment (mikluke) [inline internal/referral/rewarder.go:81]: isn't it better to use sdk.Dec?

File: internal/referral/rewarder.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+		if uPDVBalance > r.uPDVThreshold {
+			r.reward(ctx, ref, logger)
+		} else {
+			logger.Infof("balance %d less than threshold %d", uPDVBalance, r.uPDVThreshold)
+		}
+	}
+}
+
+func (r *Rewarder) reward(ctx context.Context, ref *storage.ReferralTracking, logger *log.Entry) {
```
Comment (mikluke) [inline internal/referral/rewarder.go:98]: I don't like passing logger as parameter, it's not a problem to build it here too

File: internal/referral/rewarder.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+			return fmt.Errorf("failed to transition referral to confirmed: %w", err)
+		}
+
+		if err := r.bmc.SendStakes(ref.Sender, int64(r.senderReward)); err != nil {
+			return fmt.Errorf("failed to send stakes to the referral sender: %w", err)
+		}
+
+		if err := r.bmc.SendStakes(ref.Receiver, int64(r.receiverReward)); err != nil {
+			return fmt.Errorf("failed to send stakes to the referral receiver: %w", err)
+		}
```
Comment (mikluke) [inline internal/referral/rewarder.go:111]: I think it's a good idea to send referral reward as one tx with 2 msgs

File: cmd/referral/main.go (hunk 1 above)
Comment (megaherz) [reply cmd/referral/main.go:52]: ReferralSenderReward = 10 
ReferralReceiverReward = 10
are likely to be, there is no value to make them  sdk.Coin or sdk.Int

File: internal/referral/rewarder.go (hunk 2 above)
Comment (megaherz) [reply internal/referral/rewarder.go:56]: Added

File: internal/referral/rewarder.go (hunk 3 above)
Comment (megaherz) [reply internal/referral/rewarder.go]: Indeed, added

File: internal/referral/rewarder.go (hunk 4 above)
Comment (megaherz) [reply internal/referral/rewarder.go]: created a getLogger func

File: internal/referral/rewarder.go (hunk 5 above)
Comment (megaherz) [reply internal/referral/rewarder.go]: thx, refactored

**Review by mikluke — APPROVED** [review]

**Pull request merged**